drivers/virtio/mmio/virtio_mmio.c

```diff
@@ -395,45 +395,38 @@ static struct virtio_config_ops virtio_mmio_config_ops = {
 
 static int virtio_mmio_probe(struct pf_device *pfdev)
 {
+	int rc;
 	const fdt32_t *prop;
-	int type, hwirq, prop_len;
-	int fdt_vm = pfdev->fdt_offset;
-	__u64 reg_base;
-	__u64 reg_size;
+	int prop_len, offs;
+	__u64 reg_base, reg_size;
 	void *dtb;
+	struct uk_intctlr_irq irq;
 
 	dtb = (void *)ukplat_bootinfo_get()->dtb;
-	if (fdt_vm == -FDT_ERR_NOTFOUND) {
-		uk_pr_info("device not found in fdt\n");
-		goto error_exit;
-	} else {
-		prop = fdt_getprop(dtb, fdt_vm, "interrupts", &prop_len);
-		if (!prop) {
-			uk_pr_err("irq of device not found in fdt\n");
-			goto error_exit;
-		}
+	offs = pfdev->fdt_offset;
 
-		type = fdt32_to_cpu(prop[0]);
-		hwirq = fdt32_to_cpu(prop[1]);
+	if (unlikely(offs < 0))
+		return -EINVAL;
 
-		prop = fdt_getprop(dtb, fdt_vm, "reg", &prop_len);
-		if (!prop) {
-			uk_pr_err("reg of device not found in fdt\n");
-			goto error_exit;
-		}
+	rc = uk_intctlr_fdt_get_irq(dtb, offs, 0, &irq);
+	if (unlikely(rc < 0))
+		return -EINVAL;
 
-		/* only care about base addr, ignore the size */
-		fdt_get_address(dtb, fdt_vm, 0, &reg_base, &reg_size);
-	}
+	uk_intctlr->ops->configure_irq(&irq);
```

Comment:
I would introduce a new function instead of having to access a global variable. Also I think it's sensible to have that function return an error code in case configuring the interrupt is not possible.

Reply:
> But doesn't that apply to all uk_intctlr->ops? What makes configuire_irq() different? Or are you referring to all ops?

Yes, it's about all operations. It's a bit inconsistent in the Unikraft codebase, but we do that in vfscore for example. 
It also feels a bit more consistent with the remaining library API (`uk_intctlr_irq_register`)

Reply:
Okay no problem. Would you be happy if I add macro wrappers to `uk_intctlr->ops` and leave the `uk_intctlr` global public? Alternative we can implement a stricter scheme that provides all functionality via the API.

Reply:
> I imagine the API being a set of functions whose one of the arguments is a pointer to the intctlr itself. Each driver could have its own device structure containing its correspondent intctlr pointer, which a device driver could use to call intctlr API's.

Who populates that argument and from where? And what problem does it solve?

Reply:
> > I imagine the API being a set of functions whose one of the arguments is a pointer to the intctlr itself. Each driver could have its own device structure containing its correspondent intctlr pointer, which a device driver could use to call intctlr API's.
> 
> Who populates that argument and from where? And what problem does it solve?

Well imagine it like this
- Device has irq connected to GPIO controller
- GPIO controller has its IRQ to GIC
- in DT, Device would have a phandle to the GPIO controller in its `interrupts = ` property, so the device structure could have a pointer to the GPIO controller's device structure as its irq controller
- in DT, GPIO controller would not have a phandle to GIC and we can populate that argument with the irq controller at the top of the hierarchy by default, so GPIO controller could have in its device structure a pointer to the GIC as its irq controller
- GIC would have its parent irq controller set to NULL since it is at the top.

When device wants to request an irq, the API just iteratively goes from parent to parent of the API invoking the correspondent driver method. This way you could handle nesting.

Reply:
@mschlumpp agreed, will update as suggested 👍🏼 

@mogasergiu since we don't have a use-case for nesting, let's update the API when this becomes a requirement along with any other associated changes.

Reply:
@mogasergiu feel free to mark this as resolved